rfd/0088-passwordless-windows.md

```diff
@@ -0,0 +1,216 @@
+
+### Authentication
+
+Authentication depends on server settings (otp, webauthn, passwordless).
+
+`tsh` performs the authentication process, setting all parameters for a
+passwordless / resident key via `WebAuthNAuthenticatorGetAssertion` from
+`webauthn.dll`.
+
+The way to specify if Windows Hello or FIDO device should be used is via
+mfa-mode: `tsh --mfa-mode={auto,platform,cross-platform}` - choose whether to
+use platform or cross-platform MFA. Auto defaults to platform in Windows Hello.
+
+### Detecting Windows Hello support
+
+Detecting Windows Hello and FIDO devices support is important so `tsh` may
+enable/disable related features as appropriate.
+
+Apart from Go build tags, which are a rather coarse detection mechanism, we can
+
+* Verify if `webauthn.dll` is present on system and it's version (support v1-v4)
+* Verify device supports for user-verifying platform authenticator via API call
+  `WebAuthNIsUserVerifyingPlatformAuthenticatorAvailable`
+* If device is not supporting user-verifying platform authenticator, fallback to
+  FIDO devices should be possible
+
+### `tsh` support commands
+
+The following support commands are added to `tsh` as hidden subcommands. They
+are useful to diagnose and manage certain aspects of Windows Hello support.
+
+The commands are only available on windows builds.
+
+`tsh winhello diag` - prints diagnostics about Windows Hello support (for
+example, what API version is `webauthn.dll` etc.
+
+```shell
+$ tsh winhello diag  # diag output subject to change
+> Windows version: 10
+> Signed: yes
+> Webauthn.dll version: 3
+> Supports platform user-verification: yes
+```
+
+### webauthn.dll and Teleport
+
+At the time of writing this rfd, Go wrapper around `webauthn.dll` does not
+exists. New open source repository will be created where teleport will maintain
+unofficial Go client for `webauthn.dll`.
+
+### Executable sign
+
+Although Windows allow usage of not-signed executables, it results in poor UX.
+There is warning about _using not verified app_. This can be easily fixed by
+signing the executable.
+
+### Security
+
+Security considerations are largely unchanged in relation to the
+[Passwordless RFD][passwordless rfd].
+
+[Microsoft webauthnAPI documentation][webauthn API windows docs] provides more
+information about device requirements etc.
+
+It's important to note that CLI applications with hardware key access are
+particularly dangerous, as they are not subject to the security measures
+we've come to expect from browsers (in fact, _any_ untrusted application that
+interacts with hardware keys should be used with caution).
+Those concerns are not exclusive to WebAuthn, but are made more delicate in face
+of passwordless credentials.
+
+`tsh` does its best to be well-behaved and mitigates hardware key access
+concerns by being open source, providing signed binaries and its own Cloud
+solution.
+
+### UX
+
+UX from `tsh` command perspective does not change at all in comaprison to
```

Comment:
My 2c: link to specific RFD sections and/or add a few explicit examples here. I try to write UX sections so that UX folks could read just this section as part of their review. In this case I would cover Hello/FIDO registration/authn scenarios.

It's interesting to restate behaviors like Hello-by-default, and what happens if the device is Hello-capable but the user is trying to use a FIDO fob instead.

Finally, does `tsh` remember --mfa-mode flags? It's good to note, as it makes UX a bit better.

Reply:
> Finally, does tsh remember --mfa-mode flags? It's good to note, as it makes UX a bit better.

It doesn't. I will add it to rfd and will implement it during implementation phase.

> It's interesting to restate behaviors like Hello-by-default, and what happens if the device is Hello-capable but the user is trying to use a FIDO fob instead.

So if there is nothing registered, and auto attachment is selected, windows will provide "win-hello" device registration. If you press cancel, it goes back to FIDO device (it is quite unintuitive).

During login if you have multiple devices registered, following dialog is presented and you can select security key: 
<img width="342" alt="0088-login-choice-multi" src="https://user-images.githubusercontent.com/14020794/190575824-287c8ff5-d89c-44a6-96aa-77f06df59baa.png">